File: frontend/src/mocks/schedules.ts
import type {
  Schedule,
  ScheduleConflict,
  ScheduleListQuery,
  ScheduleListResult,
  ScheduleUpsertCommand,
  ScheduleUpsertError,
  ScheduleUpsertResult,
} from '../types/home';

export const scheduleListMock: ScheduleListResult = {
  type: 'schedule.list.result',
  request_id: 'req_schedule_list_mock',
  ok: true,
  payload: {
    schedules: [
      {
        id: 'schedule_001',
        user_id: 'default_user',
        source_mode: 'voice',
        schedule_type: 'time',
        status: 'done',
        title: '复习 Java 集合基础',
        notes: '复习 List、Set 与 Map 的常见实现。',
        start_time: '2026-07-27T08:40:00+08:00',
        end_time: '2026-07-27T09:15:00+08:00',
        timezone: 'Asia/Shanghai',
        location_name: null,
        location_address: null,
        latitude: null,
        longitude: null,
        geofence_radius_meters: 100,
        geofence_armed: true,
        time_remind_offset_minutes: 15,
        time_triggered_at: '2026-07-27T08:25:00+08:00',
        geo_triggered_at: null,
        system_schedule_ref_id: 'system_schedule_001',
        system_alarm_ref_id: null,
        created_at: '2026-07-26T20:10:00+08:00',
        updated_at: '2026-07-27T09:15:00+08:00',
      },
      {
        id: 'schedule_002',
        user_id: 'default_user',
        source_mode: 'manual',
        schedule_type: 'time',
        status: 'scheduled',
        title: '需求分析讨论',
        notes: null,
        start_time: '2026-07-28T14:00:00+08:00',
        end_time: '2026-07-28T14:45:00+08:00',
        timezone: 'Asia/Shanghai',
        location_name: '张江办公室',
        location_address: '上海市浦东新区张江路 88 号',
        latitude: 31.2015,
        longitude: 121.5871,
        geofence_radius_meters: 100,
        geofence_armed: true,
        time_remind_offset_minutes: 15,
        time_triggered_at: null,
        geo_triggered_at: null,
        system_schedule_ref_id: 'system_schedule_002',
        system_alarm_ref_id: null,
        created_at: '2026-07-27T18:30:00+08:00',
        updated_at: '2026-07-27T18:30:00+08:00',
      },
      {
        id: 'schedule_003',
        user_id: 'default_user',
        source_mode: 'voice',
        schedule_type: 'time',
        status: 'done',
        title: '完成需求分析初稿',
        notes: null,
        start_time: '2026-07-29T09:25:00+08:00',
        end_time: '2026-07-29T09:50:00+08:00',
        timezone: 'Asia/Shanghai',
        location_name: null,
        location_address: null,
        latitude: null,
        longitude: null,
        geofence_radius_meters: 100,
        geofence_armed: true,
        time_remind_offset_minutes: 15,
        time_triggered_at: '2026-07-29T09:10:00+08:00',
        geo_triggered_at: null,
        system_schedule_ref_id: 'system_schedule_003',
        system_alarm_ref_id: null,
        created_at: '2026-07-28T21:30:00+08:00',
        updated_at: '2026-07-29T09:50:00+08:00',
      },
      {
        id: 'schedule_004',
        user_id: 'default_user',
        source_mode: 'manual',
        schedule_type: 'time',
        status: 'scheduled',
        title: '项目周会',
        notes: '同步本周进展与风险。',
        start_time: '2026-07-29T10:30:00+08:00',
        end_time: '2026-07-29T11:20:00+08:00',
        timezone: 'Asia/Shanghai',
        location_name: '3 号会议室',
        location_address: '上海市浦东新区张江路 88 号 5 楼',
        latitude: 31.2016,
        longitude: 121.587,
        geofence_radius_meters: 100,
        geofence_armed: true,
        time_remind_offset_minutes: 15,
        time_triggered_at: null,
        geo_triggered_at: null,
        system_schedule_ref_id: 'system_schedule_004',
        system_alarm_ref_id: null,
        created_at: '2026-07-25T10:00:00+08:00',
        updated_at: '2026-07-25T10:00:00+08:00',
      },
      {
        id: 'schedule_005',
        user_id: 'default_user',
        source_mode: 'voice',
        schedule_type: 'time',
        status: 'scheduled',
        title: '复习 JVM 内存模型',
        notes: null,
        start_time: '2026-07-29T19:30:00+08:00',
        end_time: '2026-07-29T20:05:00+08:00',
        timezone: 'Asia/Shanghai',
        location_name: null,
        location_address: null,
        latitude: null,
        longitude: null,
        geofence_radius_meters: 100,
        geofence_armed: true,
        time_remind_offset_minutes: 15,
        time_triggered_at: null,
        geo_triggered_at: null,
        system_schedule_ref_id: 'system_schedule_005',
        system_alarm_ref_id: null,
        created_at: '2026-07-28T22:00:00+08:00',
        updated_at: '2026-07-28T22:00:00+08:00',
      },
      {
        id: 'schedule_006',
        user_id: 'default_user',
        source_mode: 'manual',
        schedule_type: 'time',
        status: 'scheduled',
        title: '整理项目介绍提纲',
        notes: null,
        start_time: '2026-07-29T20:15:00+08:00',
        end_time: '2026-07-29T20:35:00+08:00',
        timezone: 'Asia/Shanghai',
        location_name: null,
        location_address: null,
        latitude: null,
        longitude: null,
        geofence_radius_meters: 100,
        geofence_armed: true,
        time_remind_offset_minutes: 15,
        time_triggered_at: null,
        geo_triggered_at: null,
        system_schedule_ref_id: 'system_schedule_006',
        system_alarm_ref_id: null,
        created_at: '2026-07-28T22:10:00+08:00',
        updated_at: '2026-07-28T22:10:00+08:00',
      },
      {
        id: 'schedule_007',
        user_id: 'default_user',
        source_mode: 'voice',
        schedule_type: 'time',
        status: 'scheduled',
        title: '整理面试问题清单',
        notes: '优先整理并发与数据库问题。',
        start_time: '2026-07-31T19:30:00+08:00',
        end_time: '2026-07-31T20:20:00+08:00',
        timezone: 'Asia/Shanghai',
        location_name: null,
        location_address: null,
        latitude: null,
        longitude: null,
        geofence_radius_meters: 100,
        geofence_armed: true,
        time_remind_offset_minutes: 15,
        time_triggered_at: null,
        geo_triggered_at: null,
        system_schedule_ref_id: 'system_schedule_007',
        system_alarm_ref_id: null,
        created_at: '2026-07-28T22:20:00+08:00',
        updated_at: '2026-07-28T22:20:00+08:00',
      },
      {
        id: 'schedule_008',
        user_id: 'default_user',
        source_mode: 'voice',
        schedule_type: 'location',
        status: 'scheduled',
        title: '到公司后提交实习简历',
        notes: '提交前再次检查附件命名。',
        start_time: null,
        end_time: null,
        timezone: null,
        location_name: '张江办公室',
        location_address: '上海市浦东新区张江路 88 号',
        latitude: 31.2015,
        longitude: 121.5871,
        geofence_radius_meters: 100,
        geofence_armed: true,
        time_remind_offset_minutes: 15,
        time_triggered_at: null,
        geo_triggered_at: null,
        system_schedule_ref_id: 'system_schedule_008',
        system_alarm_ref_id: null,
        created_at: '2026-07-29T08:00:00+08:00',
        updated_at: '2026-07-29T08:00:00+08:00',
      },
    ],
  },
};

export const scheduleListQueryMock: ScheduleListQuery = {
  type: 'schedule.list.query',
  request_id: 'req_schedule_list_001',
  payload: {
    status: null,
    include_deleted: false,
  },
};

export const scheduleUpsertMock: ScheduleUpsertCommand = {
  type: 'schedule.upsert.command',
  request_id: 'req_schedule_001',
  payload: {
    schedule_id: null,
    source_mode: 'voice',
    schedule_type: 'time',
    title: '开会',
    notes: null,
    start_time: '2026-07-29T15:00:00+08:00',
    end_time: null,
    timezone: 'Asia/Shanghai',
    location_name: '陆家嘴',
    location_address: null,
    latitude: 31.2451,
    longitude: 121.5067,
    geofence_radius_meters: 100,
    geofence_armed: true,
    time_remind_offset_minutes: 15,
  },
};

export const scheduleUpsertResultMock: ScheduleUpsertResult = {
  type: 'schedule.upsert.result',
  request_id: 'req_schedule_001',
  ok: true,
  payload: {
    schedule_id: 'schedule_001',
    schedule_type: 'time',
    status: 'scheduled',
    conflicts: [],
    geofence_armed: true,
  },
};

function findScheduleConflicts(
  command: ScheduleUpsertCommand,
  schedules: Schedule[],
  currentScheduleId: string,
): ScheduleConflict[] {
  const { end_time: endTime, start_time: startTime } = command.payload;
  if (!startTime) return [];

  const start = new Date(startTime).getTime();
  const end = endTime ? new Date(endTime).getTime() : start;
  if (!Number.isFinite(start) || !Number.isFinite(end)) return [];

  return schedules
    .filter((item) => item.id !== currentScheduleId && item.status !== 'deleted' && item.start_time)
    .filter((item) => {
      const itemStart = new Date(item.start_time!).getTime();
      const itemEnd = item.end_time ? new Date(item.end_time).getTime() : itemStart;
      return (
        Number.isFinite(itemStart) &&
        Number.isFinite(itemEnd) &&
        start <= itemEnd &&
        itemStart <= end
      );
    })
    .map((item) => ({
      schedule_id: item.id,
      title: item.title,
      start_time: item.start_time!,
      end_time: item.end_time,
    }));
}

export function createScheduleUpsertResultMock(
  command: ScheduleUpsertCommand,
  schedules: Schedule[],
  scheduleId: string,
): ScheduleUpsertResult {
  const existingSchedule = schedules.find((item) => item.id === scheduleId);

  return {
    ...scheduleUpsertResultMock,
    request_id: command.request_id,
    payload: {
      ...scheduleUpsertResultMock.payload,
      schedule_id: scheduleId,
      schedule_type: command.payload.schedule_type,
      conflicts: findScheduleConflicts(command, schedules, scheduleId),
      geofence_armed: command.payload.geofence_armed ?? existingSchedule?.geofence_armed ?? true,
    },
  };
}

export const scheduleUpsertConflictMock: ScheduleUpsertResult = {
  type: 'schedule.upsert.result',
  request_id: 'req_schedule_001',
  ok: true,
  payload: {
    schedule_id: 'schedule_001',
    schedule_type: 'time',
    status: 'scheduled',
    conflicts: [
      {
        schedule_id: 'schedule_older',
        title: '已有日程',
        start_time: '2026-07-28T15:00:00+08:00',
        end_time: '2026-07-28T16:00:00+08:00',
      },
    ],
    geofence_armed: true,
  },
};

export const scheduleUpsertErrorMock: ScheduleUpsertError = {
  type: 'schedule.upsert.error',
  request_id: 'req_schedule_001',
  ok: false,
  error: {
    code: 'VALIDATION_ERROR',
    message: '请求参数不合法',
    details: {
      field: 'schedule_type',
      reason: 'schedule_type 为 time 时 start_time 必填；为 location 时 latitude 和 longitude 必填',
    },
  },
};

File: backend/src/timeflow/infrastructure/websocket/handlers/schedules.py
"""WebSocket handlers for schedule commands and queries."""

from __future__ import annotations

import logging
from typing import Any

from pydantic import ValidationError

from timeflow.business.reminders import ReminderAudioGenerationPort
from timeflow.business.schedules import (
    ScheduleListQuery as BusinessScheduleListQuery,
)
from timeflow.business.schedules import (
    ScheduleNotFoundError,
    ScheduleRecord,
    ScheduleService,
    ScheduleValidationError,
)
from timeflow.business.schedules import ScheduleUpsertCommand as BusinessScheduleUpsertCommand
from timeflow.infrastructure.websocket.envelope import build_error_envelope
from timeflow.infrastructure.websocket.messages.envelope import ErrorDetail
from timeflow.infrastructure.websocket.messages.schedule import (
    ScheduleConflict as ScheduleConflictMessage,
)
from timeflow.infrastructure.websocket.messages.schedule import (
    ScheduleDeleted,
    ScheduleDeletedAck,
    ScheduleListQuery,
    ScheduleListResult,
    ScheduleListResultPayload,
    ScheduleSummary,
    ScheduleUpsertCommand,
    ScheduleUpsertResult,
    ScheduleUpsertResultPayload,
)
from timeflow.infrastructure.websocket.reminder_audio import ReminderAudioGenerationTracker

logger = logging.getLogger(__name__)


class ScheduleWebSocketHandlers:
    """Adapt schedule WS messages to business service calls."""

    def __init__(
        self,
        service: ScheduleService,
        reminder_audio_service: ReminderAudioGenerationPort | None = None,
        generation_tracker: ReminderAudioGenerationTracker | None = None,
    ) -> None:
        self._service = service
        self._reminder_audio_service = reminder_audio_service
        # 生成任务的注册表交给 tracker 统一持有,`ReminderAudioSender` 下发提醒前
        # 要靠它判断"这条日程的音频是不是还在生成中",所以不能是本类私有状态。
        self._generation_tracker = generation_tracker or ReminderAudioGenerationTracker(
            reminder_audio_service
        )

    async def handle_upsert(
        self,
        raw_message: dict[str, Any],
        device_id: str,
    ) -> dict[str, Any]:
        """Handle `schedule.upsert.command`."""
        del device_id
        request_id = self._extract_request_id(raw_message)
        try:
            message = ScheduleUpsertCommand.model_validate(raw_message)
            payload = message.payload
            result = self._service.upsert(
                BusinessScheduleUpsertCommand(
                    schedule_id=payload.schedule_id,
                    source_mode=payload.source_mode,
                    schedule_type=payload.schedule_type,
                    title=payload.title,
                    notes=payload.notes,
                    start_time=payload.start_time,
                    end_time=payload.end_time,
                    timezone=payload.timezone,
                    location_name=payload.location_name,
                    location_address=payload.location_address,
                    latitude=payload.latitude,
                    longitude=payload.longitude,
                    geofence_radius_meters=payload.geofence_radius_meters,
                    geofence_armed=payload.geofence_armed,
                    time_remind_offset_minutes=payload.time_remind_offset_minutes,
                )
            )
        except ValidationError as exc:
            return self._validation_error("schedule.upsert.error", request_id, exc.errors())
        except ScheduleValidationError as exc:
            return self._business_validation_error("schedule.upsert.error", request_id, exc)
        except ScheduleNotFoundError:
            return build_error_envelope(
                "schedule.upsert.error",
                request_id,
                "SCHEDULE_NOT_FOUND",
                "日程不存在",
            )

        self._submit_audio_generation(result.schedule)

        response = ScheduleUpsertResult(
            request_id=message.request_id,
            payload=ScheduleUpsertResultPayload(
                schedule_id=result.schedule_id,
                schedule_type=result.schedule_type,
                status=result.status,
                conflicts=[
                    ScheduleConflictMessage(
                        schedule_id=conflict.schedule_id,
                        title=conflict.title,
                        start_time=conflict.start_time,
                        end_time=conflict.end_time,
                    )
                    for conflict in result.conflicts
                ],
                geofence_armed=result.geofence_armed,
            ),
        )
        return response.model_dump()

    async def aclose(self) -> None:
        """Cancel pending TTS generation tasks during application shutdown."""
        await self._generation_tracker.aclose()

    async def _delete_audio(self, schedule_id: str) -> None:
        if self._reminder_audio_service is None:
            return
        await self._generation_tracker.discard(schedule_id)
        await self._reminder_audio_service.delete(schedule_id)

    def _submit_audio_generation(self, schedule: ScheduleRecord) -> None:
        self._generation_tracker.submit(schedule)

    async def handle_list(
        self,
        raw_message: dict[str, Any],
        device_id: str,
    ) -> dict[str, Any]:
        """Handle `schedule.list.query`."""
        del device_id
        request_id = self._extract_request_id(raw_message)
        try:
            message = ScheduleListQuery.model_validate(raw_message)
            result = self._service.list(
                BusinessScheduleListQuery(
                    status=message.payload.status,
                    include_deleted=message.payload.include_deleted,
                )
            )
        except ValidationError as exc:
            return self._validation_error("schedule.list.error", request_id, exc.errors())
        except ScheduleValidationError as exc:
            return self._business_validation_error("schedule.list.error", request_id, exc)

        response = ScheduleListResult(
            request_id=message.request_id,
            payload=ScheduleListResultPayload(
                schedules=[
                    ScheduleSummary(
                        id=schedule.id,
                        user_id=schedule.user_id,
                        source_mode=schedule.source_mode,
                        schedule_type=schedule.schedule_type,
                        status=schedule.status,
                        title=schedule.title,
                        notes=schedule.notes,
                        start_time=schedule.start_time,
                        end_time=schedule.end_time,
                        timezone=schedule.timezone,
                        location_name=schedule.location_name,
                        location_address=schedule.location_address,
                        latitude=schedule.latitude,
                        longitude=schedule.longitude,
                        geofence_radius_meters=schedule.geofence_radius_meters,
                        geofence_armed=schedule.geofence_armed,
                        time_remind_offset_minutes=schedule.time_remind_offset_minutes,
                        time_triggered_at=schedule.time_triggered_at,
                        geo_triggered_at=schedule.geo_triggered_at,
                        system_schedule_ref_id=schedule.system_schedule_ref_id,
                        system_alarm_ref_id=schedule.system_alarm_ref_id,
                        created_at=schedule.created_at,
                        updated_at=schedule.updated_at,
                    )
                    for schedule in result.schedules
                ]
            ),
        )
        return response.model_dump()

    async def handle_deleted(
        self,
        raw_message: dict[str, Any],
        device_id: str,
    ) -> dict[str, Any]:
        """Handle `schedule.deleted`。成功/失败共用同一个响应类型,靠 `ok` 区分,
        不像 upsert/list 那样走 request_id 关联的 result/error 分支。"""
        del device_id
        try:
            message = ScheduleDeleted.model_validate(raw_message)
        except ValidationError:
            raw_schedule_id = raw_message.get("schedule_id")
            return ScheduleDeletedAck(
                schedule_id=raw_schedule_id if isinstance(raw_schedule_id, str) else "",
                ok=False,
                error=ErrorDetail(code="VALIDATION_ERROR", message="请求参数不合法"),
            ).model_dump(exclude_none=True)

        try:
            self._service.delete(message.schedule_id)
        except ScheduleNotFoundError:
            return ScheduleDeletedAck(
                schedule_id=message.schedule_id,
                ok=False,
                error=ErrorDetail(
                    code="SCHEDULE_DELETE_FAILED",
                    message="日程删除失败",
                    details={"reason": "schedule_not_found"},
                ),
            ).model_dump(exclude_none=True)
        except ScheduleValidationError as exc:
            return ScheduleDeletedAck(
                schedule_id=message.schedule_id,
                ok=False,
                error=ErrorDetail(
                    code="VALIDATION_ERROR",
                    message="请求参数不合法",
                    details={"field": exc.field, "reason": exc.reason},
                ),
            ).model_dump(exclude_none=True)

        await self._delete_audio(message.schedule_id)
        return ScheduleDeletedAck(schedule_id=message.schedule_id, ok=True).model_dump(
            exclude_none=True
        )

    @staticmethod
    def _extract_request_id(raw_message: dict[str, Any]) -> str | None:
        value = raw_message.get("request_id")
        return value if isinstance(value, str) else None

    @staticmethod
    def _validation_error(
        message_type: str,
        request_id: str | None,
        errors: list[Any],
    ) -> dict[str, Any]:
        return build_error_envelope(
            message_type,
            request_id,
            "VALIDATION_ERROR",
            "请求参数不合法",
            {"errors": errors},
        )

    @staticmethod
    def _business_validation_error(
        message_type: str,
        request_id: str | None,
        exc: ScheduleValidationError,
    ) -> dict[str, Any]:
        return build_error_envelope(
            message_type,
            request_id,
            "VALIDATION_ERROR",
            "请求参数不合法",
            {"field": exc.field, "reason": exc.reason},
        )


__all__ = ["ScheduleWebSocketHandlers"]
